matrices: Improved speed of solve()

friyaz opened the pull request:
matrices : increase speed of solve()
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed

Fixes issue [#15178](https://github.com/sympy/sympy/issues/15178) .
Improved speed of solve() by using Gauss Jordan elimination
method when method choosen is Gauss Elimination . This inc-
reases the speed as inverse is not calculated .


#### Other comments

#### Release Notes



<!-- BEGIN RELEASE NOTES -->
* matrices
  * improve speed of `.solve()` in `matrices.py`
<!-- END RELEASE NOTES -->

friyaz committed:
matrices: Improved speed of solve()

Improved speed of solve() by using Gauss Jordan elimination method. Earlier solve() used to evaluate inverse which was slow.

friyaz committed:
matrices: Improve speed of solve()

Fixes issue #15178 .
Improved speed of solve() by using Gauss Jordan elimination method.
Earlier solve() used to evaluate inverse which was slow.I have modifi-
ed it to use Gauss Jordan elimination method which will be fast.
Since now slove does not need the argument method so i have remo-
ved method from argument list.

moorepants reviewed (dismissed):
Thanks for the update. I believe changing `.solve()` to a more efficient method has been discussed before but we do need to maintain the API, in general. See https://github.com/sympy/sympy/wiki/Deprecating-policy

If you support all existing methods but make use of a more efficient solver as a default, you may get consensus to merge the change.

friyaz committed:
matrices : increase speed of solve()
References to other Issues or PRs
Brief description of what is fixed or changed

Fixes issue #15178 .
Improved speed of solve() by using Gauss Jordan elimination
method when method choosen is Gauss Elimination . This inc-
reases the speed as inverse is not calculated .
Other comments
Release Notes

    matrices
        increase speed of solve()

friyaz committed:
corrected indentation

FrackeR011 commented:
I have done few changes . i dont know whether they are enough .

friyaz committed:
corrected indentation

friyaz committed:
correcting format

friyaz committed:
correcting format again

friyaz committed:
correcting format again

friyaz committed:
correcting format again

moorepants commented:
I'm wondering if there was any difference in Gaussian Elimination and Gauss Jordan Elimination. Here you are switching form one to another, but I'm not sure if we actually distinguish in any way in the existing code. Can you comment on that? I don't really remember the fine details of the difference. The wikipedia article does allude to computational advantages of one over the other.

FrackeR011 commented:
The problem which understood was that instead of calculating inverse and then multiplying it with other matrix , `.solve()` should calculate rref of augmented matrix as mentioned in [#15178](https://github.com/sympy/sympy/issues/15178) . I think i have done something else than was asked . @isuruf  can oyu help me with this one ?

isuruf commented:
I'm not sure why people would be relying on calculating the inverse and multiplying. `solve` to me should be something that users can use for solving a system efficiently and not care about details just like `np.linalg.solve`. For a more concerned user, they can use the `method` argument to specify which method to use and `solve` would be a general interface to all the methods that sympy have.

moorepants commented:
I agree with you @isuruf, but I'm more curious about what the author of `.solve()` intended it to be and what users have actually used it for, from a backwards compatibility perspective. Are we changing `.solve()` to be what you suggest or is it already that. It isn't completely clear to me what it is.

moorepants commented:
But if no results of `.solve()` change if we change the internals and it only gets faster and more efficient, I guess what I'm talking about is moot.

moorepants commented:
It looks like the `.solve()` method was added in 0.7.3 with this commit: https://github.com/sympy/sympy/commit/a59b3128c91dbcfdde4f24c93bf95232b861251e as part of https://github.com/sympy/sympy/pull/1535, which is a huge PR. It isn't clear why `.solve()` was defined like it is, but @smichr made the changes back then. @smichr could you comment on what `Matrix.solve()` is intended to be? Would there be any reasons not to switch from calculating the inverse explicitly to more efficient computational methods? The docstring says that solve "Return solution to self*soln = rhs using given inversion method." So it seems you intended it to use the inversion method.

isuruf commented:
ping @smichr

smichr commented:
> could you comment on what `Matrix.solve()` is intended to be

I think I took over the changes initiated by the then-stale #525. In the commit string of the cited commit I said, 

"These methods default to different methods than were given to the
Sparse versions of the same, following the preferences of inversion."

So I assume that I was just following what was being used by `inv`. I don't see any reason not to make `solve` as efficient as possible. I do recall, however, spending some significant time making sure that the small-n case (like up to 3?) was solved as quickly as possible. So whatever you do, check that the speed of small systems doesn't regress if a change in the default method is used.

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -4020,6 +4020,9 @@ def solve(self, rhs, method='GE'):
         """Return solution to self*soln = rhs using given inversion method.
```
How about changing the docstring and behavior as follows:

```
"""Return the unique soln making self*soln = rhs.

If there is not a unique solution then a ValueError will be raised. If `self` is not
square, a ValueError and a different routine for solving the system will be suggested.

When the method is GJ, the Gauss-Jordan elimination will be used. To use a different
method and to compute the solution via the inverse, use a method defined in the
.inv() docstring.
"""

The `method` will default to 'GJ' but `raise ValueError("Matrix det == 0; not invertible. Try gauss_jordan_solve to obtain a parametric solution.")` if a parameter matrix is returned.

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -4029,6 +4032,8 @@ def solve(self, rhs, method='GE'):
             elif self.rows > self.cols:
                 raise ValueError('For over-determined system, M, having '
                                  'more rows than columns, try M.solve_least_squares(rhs).')
+        if method == 'GE':
```
```
if method == 'GJ':
    soln, param = self.gauss_jordan_solve(rhs)
    if param:
        raise ValueError ( as given in above comment )
    return soln
else:
```

smichr replied:
This would then need a test like
```
A = Matrix([[1, 2], [2, 4]])
b = Matrix([[ 3], [-3]])
raises(ValueError, lambda: A.solve(b))
A = Matrix([[1, 2], [2, 4]])
b = Matrix([[ 3], [6]])
raises(ValueError, lambda: A.solve(b))
```
to confirm that even when GJ is used, ValueErrors are raised for non-invertible systems (which is what the current solve implicitly does).

FrackeR011 replied:
Thanks @smichr for the review . I agree with your changes . Should i commit the change or wait for others approval as well .

smichr replied:
I recommend commiting the changes. This is the only way to preserve the current behavior of the function. Because it is currently based on `inv` it raises an error when the matrix is singular. G-J won't do that so the changes I recommend catch that case and raise an error while still using a faster routine to calculate the solution.

friyaz committed:
matrices.py : modified implementation of solve()

Modified solve() to preserve current behaviour while impreving performance .
smichr code is ued .

friyaz committed:
Merge remote-tracking branch 'upstream/master' into 15178

friyaz committed:
matrices.py : changed implementation of solve()

Performance of solve() is improved while retaining current behaviour .

friyaz committed:
added test for solve()

FrackeR011 commented:
@moorepants please have a look now .

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -4061,21 +4061,29 @@ def solve_least_squares(self, rhs, method='CH'):
+            return soln
         else:
-            return self.inv(method=method) * rhs
+             return self.inv(method=method)*rhs
```
dedent by one space here

friyaz committed:
fixed indentation
